### DISCUSSION TITLE
How to give permissions to dependency package?

### DISCUSSION THREAD

**Discussion by FreePhoenix888** [deep-foundation/Discussions#11] [Q&A] [answered]
# Abstract question
I have package `A` that has joins to admin and packages.
`A` package depends on `B` package
How to make `B` to have the same permissions as `A` package has?
I have tried to create join from `A` to `B` but it does not help

# Not abstract question
I have `@deep-foundation/deep-memo` package.  
`@deep-foundation/deep-memo` depends on `@deep-foundation/device` package.  
`@deep-foundation/device` depends on `@freephoenix888/object-to-links-async-converter`  

`@freephoenix888/object-to-links-async-converter` is unable to create links even if there are 
- a Join link from deep-memo to admin and packages
- a Join link from deep-memo to `@deep-foundation/device`
- a Join link from `@deep-foundation/device` to `@freephoenix888/object-to-links-async-converter`

**Comment by konard** [comment]
1. Why it is not on stackoverflow with reproducible example?
2. Why you cannot just create direct join from each package to admin?

**Reply by FreePhoenix888** [reply]
> Why it is not on stackoverflow with reproducible example?

Because I have written one good Question and Answer but no one from deep supported one expect you therefore that question was closed. I am not going to publish something about deep on stackoverflow until we get enough people that are able and want to support us there

> Why you cannot just create direct join from each package to admin?  

I do not like this solution. There are a lot of packages

**Comment by FreePhoenix888** [comment]
I have created Join from `deep-memo` to admin and packages
I have created Join from `deep-memo` to `device`
I have created Join from device to `@freephoenix888/object-to-links-async-converter`

Selects:
![image](https://github.com/deep-foundation/Discussions/assets/66206278/ce745649-b9bc-4b60-8f8c-c657ff417949)

There is a handler from `@freephoenix888/object-to-links-async-converter`. `device` has `HandleUpdate` from `Device` to that handler. Error when updating Device:
```json
{
  "name": "ApolloError",
  "stack": "ApolloError: check constraint of an insert/update permission has failed\n    at new ApolloError (/node_modules/@apollo/client/errors/errors.cjs:34:28)\n    at /node_modules/@apollo/client/core/core.cjs:1289:47\n    at both (/node_modules/@apollo/client/utilities/utilities.cjs:997:53)\n    at /node_modules/@apollo/client/utilities/utilities.cjs:990:72\n    at new Promise (<anonymous>)\n    at Object.then (/node_modules/@apollo/client/utilities/utilities.cjs:990:24)\n    at Object.next (/node_modules/@apollo/client/utilities/utilities.cjs:998:49)\n    at notifySubscription (/node_modules/zen-observable/lib/Observable.js:135:18)\n    at onNotify (/node_modules/zen-observable/lib/Observable.js:179:3)\n    at SubscriptionObserver.next (/node_modules/zen-observable/lib/Observable.js:235:7)",
  "message": "check constraint of an insert/update permission has failed",
  "clientErrors": [],
  "networkError": null,
  "graphQLErrors": [
    {
      "message": "check constraint of an insert/update permission has failed"
    }
  ]
}
```

**Comment by FreePhoenix888** [comment]
![image](https://github.com/deep-foundation/Discussions/assets/66206278/bfea0185-e3f5-477c-a333-3688219198f0)
If I directly give admin and packages permissions to device - it does not help
If I directly give admin and packages permissions to object-to-links-converter - it does help

Still we have a lot of packages and their dependencies and if we say our rules are "flexible" they must be flexible, are not they?

**Comment by FreePhoenix888** [comment, marked as answer]
My problem was existing because I have mixed up from_id and to_id. Here is what I had to use:
```ts
const joinTypeLinkId = await deep.id("@deep-foundation/core", "Join");
const packageLinkId = await deep.id("@freephoenix888/object-to-links-async-converter");
await deep.insert([
  {
    type_id: joinTypeLinkId,
    from_id: packageLinkId,
    to_id: await deep.id("@deep-foundation/capacitor-device"),
  },
])
```
```ts
const joinTypeLinkId = await deep.id("@deep-foundation/core", "Join");
const packageLinkId = await deep.id("@deep-foundation/capacitor-device");
await deep.insert([
  {
    type_id: joinTypeLinkId,
    from_id: packageLinkId,
    to_id: await deep.id("@deep-foundation/deep-memo"),
  },
])
```
```ts
const joinTypeLinkId = await deep.id("@deep-foundation/core", "Join");
const packageLinkId = await deep.id("@deep-foundation/deep-memo");
await deep.insert([
  {
    type_id: joinTypeLinkId,
    from_id: packageLinkId,
    to_id: await deep.id('deep', 'users', 'packages'),
  },
  {
    type_id: joinTypeLinkId,
    from_id: packageLinkId,
    to_id: await deep.id('deep', 'admin'),
  },
])
```

Join's `from` must be the package that need permissions of `to`
